File: app/core/registry.py
_REGISTRY: dict = {}


class RegisteredHandler:
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        if hasattr(cls, "name"):
            _REGISTRY[cls.name] = cls

    def execute(self):
        raise NotImplementedError


class EmailHandler(RegisteredHandler):
    name = "email"

    def execute(self):
        print("Sending email notification...")


class SlackAlertHandler(RegisteredHandler):
    name = "slack"

    def execute(self):
        print("Posting alert to Slack...")


def get_handler(name: str) -> RegisteredHandler:
    cls = _REGISTRY.get(name)
    if cls is None:
        raise KeyError(f"No handler registered for: {name}")
    return cls()

File: web/src/core/registry.ts
const _REGISTRY: Record<string, new () => RegisteredHandler> = {};

export function RegisterHandler(name: string) {
  return function <T extends new (...args: any[]) => RegisteredHandler>(constructor: T) {
    _REGISTRY[name] = constructor;
    return constructor;
  };
}

export abstract class RegisteredHandler {
  abstract execute(): void;
}

@RegisterHandler("email")
export class EmailHandler extends RegisteredHandler {
  execute(): void {
    console.log("Sending email notification...");
  }
}

@RegisterHandler("slack")
export class SlackAlertHandler extends RegisteredHandler {
  execute(): void {
    console.log("Posting alert to Slack...");
  }
}

export function getHandler(name: string): RegisteredHandler {
  const cls = _REGISTRY[name];
  if (!cls) {
    throw new Error(`No handler registered for: ${name}`);
  }
  return new cls();
}
